glib/src/gstring.rs

```diff
@@ -678,50 +678,48 @@ impl From<&GStr> for GString {
     }
 }
 
-impl From<&str> for GString {
+impl TryFrom<&str> for GString {
+    type Error = std::ffi::NulError;
     #[inline]
-    fn from(s: &str) -> Self {
-        // Allocates with the GLib allocator
-        unsafe {
-            // No check for valid UTF-8 here
-            let copy = ffi::g_malloc(s.len() + 1) as *mut c_char;
-            ptr::copy_nonoverlapping(s.as_ptr() as *const c_char, copy, s.len() + 1);
-            ptr::write(copy.add(s.len()), 0);
-
-            GString(Inner::Foreign {
-                ptr: ptr::NonNull::new_unchecked(copy),
-                len: s.len(),
-            })
-        }
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        s.to_owned().try_into()
```

Comment:
This now creates a `CString`-based `GString`, so doesn't use the GLib allocator anymore (also in some other places). The whole point here was that if we have to copy anyway then we'd use the GLib allocator.

The idea was that `GString` will contain a string allocated with the GLib allocator whenever a copy has to happen and only otherwise contain something based on the `std` allocator. If that makes sense, I don't know and this needs some more thinking.

The goal would be to reduce the number of copies as much as possible, and e.g. later being able to pass a `GString` to a `transfer full` C function without any copies (and a `&GStr` to a `transfer none` C function).

Reply:
That makes sense, I've changed it back and updated it to check for interior nuls